src/geometry.h

```diff
@@ -0,0 +1,1104 @@
+    }
+
+    bool operator==(const Bbox3<T> other) const {
+        return p_min == other.p_min && p_max == other.p_max;
+    }
+
+    bool operator!=(const Bbox3<T> other) const { return !(*this == other); }
+
+    Vector3<T> diagonal() const { return p_max - p_min; }
+
+    T surface_area() const {
+        Vector3<T> d = diagonal();
+        return 2 * (d.x * d.y + d.x * d.z + d.y * d.z);
+    }
+
+    T volume() const {
+        Vector3<T> d = diagonal();
+        return d.x * d.y * d.z;
+    }
+
+    Point3<T> lerp(const Point3f& t) const {
+        return {static_cast<T>(turner::lerp(t.x, p_min.x, p_max.x)),
+                static_cast<T>(turner::lerp(t.y, p_min.y, p_max.y)),
+                static_cast<T>(turner::lerp(t.z, p_min.z, p_max.z))};
+    }
+
+    std::pair<Bbox3<T>, Bbox3<T>> split(size_t plane_ax,
+                                        float plane_pos) const {
+        assert(plane_ax < 3);
+        assert(p_min[plane_ax] - EPS <= plane_pos);
+        assert(plane_pos <= p_max[plane_ax] + EPS);
+
+        Point3<T> lmax = p_max;
+        lmax[plane_ax] = static_cast<T>(plane_pos);
+        Point3<T> rmin = p_min;
+        rmin[plane_ax] = plane_pos;
+
+        return {{p_min, lmax}, {rmin, p_max}};
+    }
+
+    std::string to_string() const {
+        return "[" + p_min.to_string() + ", " + p_max.to_string() + "]";
+    }
+
+public:
+    Point3<T> p_min;
+    Point3<T> p_max;
+};
+
+using Bbox3f = Bbox3<float>;
+using Bbox3i = Bbox3<int>;
+
+template <typename T>
+Bbox3<T> bbox_union(const Bbox3<T>& b, const Point3<T>& p) {
+    return {min(b.p_min, p), max(b.p_max, p)};
+}
+
+template <typename T>
+Bbox3<T> bbox_union(const Bbox3<T>& b1, const Bbox3<T>& b2) {
+    return {min(b1.p_min, b2.p_min), max(b1.p_max, b2.p_max)};
+}
+
+template <typename T>
+Bbox3<T> intersect(const Bbox3<T>& b1, const Bbox3<T>& b2) {
+    return {max(b1.p_min, b2.p_min), min(b1.p_max, b2.p_max)};
+}
+
+template <typename T> bool overlaps(const Bbox3<T>& b1, const Bbox3<T>& b2) {
+    bool x = (b1.p_max.x >= b2.p_min.x) && (b1.p_min.x <= b2.p_max.x);
+    bool y = (b1.p_max.y >= b2.p_min.y) && (b1.p_min.y <= b2.p_max.y);
+    bool z = (b1.p_max.z >= b2.p_min.z) && (b1.p_min.z <= b2.p_max.z);
+    return x && y && z;
+}
+
+template <typename T> bool inside(const Point3<T>& p, const Bbox3<T>& b) {
+    return (p.x >= b.p_min.x && p.x <= b.p_max.x && p.y >= b.p_min.y &&
+            p.y <= b.p_max.y && p.z >= b.p_min.z && p.z <= b.p_max.z);
+}
+
+//
+// Explicit template initialization
+//
+
+template class Vector2<float>;
+template class Vector2<int>;
+template class Vector3<float>;
+template class Vector3<int>;
+
+template class Point2<float>;
+template class Point2<int>;
+template class Point3<float>;
+template class Point3<int>;
+
+template class Normal3<float>;
+
+template class Bbox2<float>;
+template class Bbox2<int>;
+template class Bbox3<float>;
+template class Bbox3<int>;
+
+} // namespace turner
```

Comment:
I took the implementation mostly from the book. I would prefer not to do it, since a lot of stuff is copy pasted. You change one operator, you have to change it most likely in 3 other places, and for that I would like to have it in one place.

I also do not see any benefits of having separate files. Especially, since you always want to include the stuff here altogether.